Line/iOS/LINELoginDemo/LINELoginDemo/LineAuthManager.m

```diff
@@ -0,0 +1,230 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#import "LineAuthManager.h"
+#import <LineAdapter/LineSDK.h>
+#import <GTMHTTPFetcher.h>
+#import "Constants.h"
+@import FirebaseAuth;
+
+@interface LineAuthManager ()
+
+@property (nonatomic, strong) LineAdapter *lineAdapter;
+@property (nonatomic, weak) UIViewController *topViewController;
+@property (nonatomic, strong) FIRAuthResultCallback resultCallback;
+
+@property (nonatomic, strong) GTMHTTPFetcher *firebaseTokenFetcher;
+
+@end
+
+@implementation LineAuthManager
+
+#pragma mark - Step 1: Communication with LINE SDK
+- (void)startLINELoginWithTopViewController:(UIViewController *)viewController completionHandler:(FIRAuthResultCallback)callback {
+    self.topViewController = viewController;
+    self.resultCallback = callback;
+    
+    // STEP 1: User logins with LINE and get their LINE access token
+    if ([self.lineAdapter isAuthorized]) {
+        // If the authentication and authorization process has already been performed, start step 2 validating LINE access token
+        NSString *lineAccessToken = self.lineAdapter.getLineApiClient.accessToken;
+        [self requestFirebaseAuthTokenWithLINEAccessToken:lineAccessToken];
+    } else {
+        if ([self.lineAdapter canAuthorizeUsingLineApp]) {
+            // Authenticate with LINE application
+            [self.lineAdapter authorize];
+        }
+        else {
+            // Authenticate with WebView
+            UIViewController *viewController;
+            viewController = [[LineAdapterWebViewController alloc] initWithAdapter:self.lineAdapter
+                                                            withWebViewOrientation:LineAdapterWebViewOrientationAll];
+            [[viewController navigationItem] setLeftBarButtonItem:[LineAdapterNavigationController
+                                                                   barButtonItemWithTitle:@"Cancel" target:self action:@selector(cancel:)]];
+            UIViewController *navigationController;
+            navigationController = [[LineAdapterNavigationController alloc] initWithRootViewController:viewController];
+            [self.topViewController presentViewController:navigationController animated:YES completion:nil];
+        }
+    }
+}
+
+- (void)cancel:(id)sender {
+    [self.topViewController dismissViewControllerAnimated:YES completion:nil];
+}
+
+- (void)lineAdapterAuthorizationDidChange:(NSNotification*)aNotification {
+    LineAdapter *_adapter = [aNotification object];
```

Comment:
Fixed in this [commit](https://github.com/firebase/custom-auth-samples/pull/1/commits/cfb5c9d622d615a16b4c27754be32543065571cc) :)
My careless mistake when reusing code from LINE SDK official sample code